docs_nnx/key_concepts.md

```diff
@@ -0,0 +1,320 @@
+linear = nnx.Linear(in_features=4, out_features=2, rngs=nnx.Rngs(42))
+
+# Flax created a `Param` wrapper over the actual `jax.Array` parameter to track metadata
+print(type(linear.kernel))        # flax.nnx.Param
+print(type(linear.kernel.value))  # jax.Array
+
+# The computation of the two are the same
+x = jax.random.normal(jax.random.key(0), (2, 4))
+flax_y = linear(x)
+jax_y = jax_linear(x, linear.kernel.value, linear.bias.value)
+assert jnp.array_equal(flax_y, jax_y)
+```
+
+    <class 'flax.nnx.variablelib.Param'>
+    <class 'jaxlib._jax.ArrayImpl'>
+
+
+## Pytrees
+
+Your code likely needs more than one `jax.Array`. A **pytree** is a container structure of multiple pytrees, possibly nested. It is a key and handly concept in the JAX world.
+
+Many things are pytrees: Python dicts, lists, tuples, dataclasses, and more. The key is that a pytree can be "flattened" into multiple children, which are either pytrees or individual `jax.Array`s. JAX allows you to  register your own classes as pytree nodes.
+
+Pytree is the primary data holder in JAX. When JAX transforms see a pytree argument, they automatically trace its internal `jax.Array`s when compiling. Therefore, it's crucial to organize your data as pytrees. [JAX pytree documentation](https://docs.jax.dev/en/latest/working-with-pytrees.html) has a thorough overview on pytrees and JAX APIs to manipulate them.
+
+In Flax, a `Module` is not a pytree, but its state is a pytree. This how we implement Flax transforms (e.g. `flax.nnx.vmap` instead of `jax.vmap`) - the model is first converted to a JAX-compatible pytree dict, then the underlying JAX transforms are called.
+
+In the future, Flax Modules will also become pytrees, allowing us to directly using JAX transforms.
+
+
+```python
+state = nnx.state(linear)
+arrays, _ = jax.tree.flatten_with_path(state)
+assert len(arrays) > 1
+for kp, x in arrays:
+  print(f'{jax.tree_util.keystr(kp)}: {x}')
+```
+
+    ['bias'].value: [0. 0.]
+    ['kernel'].value: [[ 0.04119061 -0.2629074 ]
+     [ 0.6772455   0.2807398 ]
+     [ 0.16276604  0.16813846]
+     [ 0.310975   -0.43336964]]
+
+
+
+```python
+# `nnx.Module` is not a pytree (yet). Therefore, flattening it will only return itself.
+arrays, _ = jax.tree.flatten(linear)
+assert len(arrays) == 1
+assert arrays[0] == linear
+```
+
+## Traced vs. static data
+
+A pytree *contains* JAX arrays, but a pytree is *more than* its JAX arrays. For example, a dictionary keeps information like the key of every array, and it might contain entries that are not JAX arrays. From JAX's standpoint, all data are one of the two types:
+
+* **Traced** ("dynamic") data: JAX will trace them during compilation and optimize the operations upon them. If they stay inside a pytree argument, `jax.tree.flatten` must return them as leaves. They must be data values (`jax.Array`, Numpy array, scalar, etc).
+
+* **"Static"** data: They stay as simple Python objects that don't get traced by JAX.
```

Comment:
Would be good to mention that they must implement `__eq__` and `__hash__`.

Reply:
Done.